File: html/old/R2/SACO R2 2009 (Solutions)/ariprog.java
// SACO Second Round 2009
// Problem: Arithmetic Progression
// Author:  Julian Kenwood

import java.util.*;
class ariprog {
    public static void main(String args[]) {
        Scanner s = new Scanner(System.in);
        int[] seq = new int[100];
        int pos = 0;
        while(s.hasNextInt())
            seq[pos++] = s.nextInt();
        int best_len = 0, best_a = 0, best_d = 0, best_num = 0;
        for(int i = 0; i < pos; ++i) {
            for(int j = 0; j < pos; ++j) {
                int delta = seq[j] - seq[i];
                if(delta <= 0) continue;
                int cur = seq[j] + delta, cur_len = 2;
                while(true) {
                    boolean found = false;
                    for(int k = 0; k < pos; ++k)
                        if(seq[k] == cur) {
                            found = true;
                            break;
                        }
                    if(!found)
                        break;
                    cur_len++;
                    cur += delta;
                }
                if(cur_len > best_len) {
                    best_len = cur_len;
                    best_a = seq[i];
                    best_d = delta;
                    best_num = 1;
                } else if(cur_len == best_len) {
                    if(best_a > seq[i]) {
                        best_a = seq[i];
                        best_d = delta;
                    } else if(best_a == seq[i]) {
                        best_d = best_d > delta ? delta : best_d;
                    }
                    best_num++;
                }
            }
        }
        System.out.println("Arithmetic Progression Length = " + String.valueOf(best_len));
        System.out.println("Number of Arithmetic Progressions = " + String.valueOf(best_num));
        System.out.println("Arithmetic Progression First Term = " + String.valueOf(best_a));
        System.out.println("Arithmetic Progression Common Difference = " + String.valueOf(best_d));
    }
}

File: html/old/R2/SACO R2 2009 (Solutions)/ariprog.py
#!/usr/bin/env python
# SACO Second Round 2009
# Problem: Arithmetic Progression
# Author:  Marco Gallotta

N = int(raw_input("Enter N: "))
num = [int(raw_input("Enter a number: ")) for i in xrange(N)]
num.sort()
longest, count = [], 0
for i, first in enumerate(num):
  for j, second in enumerate(num[i+1:]):
    d = second - first
    seq = [first, second]
    for n in num[j+1:]:
      if n == first + d * len(seq):
        seq.append(n)
    if len(seq) > len(longest):
      longest, count = seq, 1
    elif len(seq) == len(longest):
      count += 1
print "Arithmetic Progression Length =",            len(longest)
print "Number of Arithmetic Progressions =",        count
print "Arithmetic Progression First Term =",        longest[0]
print "Arithmetic Progression Common Difference =", longest[1] - longest[0]
